compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp

```diff
@@ -182,13 +182,70 @@ static LogicalResult copyToWorkgroupMemory(OpBuilder &b, Value src, Value dst) {
   return success();
 }
 
+/// Returns the indices of the transposed operands in a linalg generic.
+static SmallVector<int64_t> getTransposedOperands(linalg::GenericOp linalgOp) {
+  // Determine which operands to promote:
+  SmallVector<int64_t> transposedOperands;
+  for (auto indexValue : llvm::enumerate(linalgOp.getIndexingMapsArray())) {
+    int64_t opIndex = indexValue.index();
+    auto indexMap = indexValue.value();
+    if (!indexMap.isEmpty() && indexMap.isPermutation()) {
+      // Ensure that the fasted moving dimension (the last one) is permuted
+      // otherwise data isn't moved.
+      if (indexMap.getDimPosition(indexMap.getNumDims() - 1) !=
+          indexMap.getNumDims() - 1) {
+        // Add operand to promote to list and mark the linalg for this
+        // promotion.
+        transposedOperands.push_back(opIndex);
+      }
+    }
+  }
+  return transposedOperands;
+}
+
+using PromotionFilterFunction = std::function<LogicalResult(Operation *op)>;
+
+/// Returns true if op is appropriate transpose for promotion.
+static LogicalResult transposeFilter(
+    Operation *op, ArrayRef<int64_t> filterOperandsToPromote) {
+  auto linalgOp = dyn_cast<linalg::GenericOp>(op);
+  if (!linalgOp) return failure();
+
+  if (linalgOp.getNumParallelLoops() < 2) {
+    return failure();
+  }
+
+  // Ensure the operands to promote matches this filter.
+  SmallVector<int64_t> operandsToPromote = getTransposedOperands(linalgOp);
+  if (operandsToPromote.size() != filterOperandsToPromote.size()) {
+    return failure();
+  }
+  if (operandsToPromote != filterOperandsToPromote) {
+    return failure();
+  } else {
+    return success();
+  }
+}
```

Comment:
I still feel like this code is unnecessary, it could just compare `op` to the operation we used to create `filterOperandsToPromote` instead.